File: lachancuoicgoi_flutter/lib/analysis/l3/core/gemini_metrics.dart
class GeminiMetrics {
  GeminiMetrics._();

  /// Singleton instance used throughout the app.
  static final GeminiMetrics instance = GeminiMetrics._();

  int _totalCalls = 0;
  int _successCalls = 0;
  int _failureCalls = 0;
  int _cacheHits = 0;
  int _cacheMisses = 0;
  int _totalLatencyMs = 0;
  final Map<int, int> _callsPerKey = <int, int>{};
  final Map<int, int> _errorsPerKey = <int, int>{};
  final List<String> _errorLogs = <String>[];

  List<String> get recentErrors => List<String>.unmodifiable(_errorLogs);

  void addErrorLog(String log) {
    _errorLogs.add(log);
    if (_errorLogs.length > 100) {
      _errorLogs.removeAt(0);
    }
  }

  void recordCall({
    required bool success,
    required int latencyMs,
    int keyIndex = -1,
  }) {
    _totalCalls++;
    if (success) {
      _successCalls++;
    } else {
      _failureCalls++;
      if (keyIndex >= 0) {
        _errorsPerKey[keyIndex] = (_errorsPerKey[keyIndex] ?? 0) + 1;
      }
    }
    _totalLatencyMs += latencyMs;
    if (keyIndex >= 0) {
      _callsPerKey[keyIndex] = (_callsPerKey[keyIndex] ?? 0) + 1;
    }
  }

  void recordCacheHit() {
    _cacheHits++;
  }

  void recordCacheMiss() {
    _cacheMisses++;
  }

  MetricsSnapshot getSnapshot() {
    final totalRequests = _cacheHits + _cacheMisses;
    final summaries =
        _callsPerKey.entries
            .map(
              (entry) => KeyMetricSummary(
                index: entry.key,
                calls: entry.value,
                errors: _errorsPerKey[entry.key] ?? 0,
              ),
            )
            .toList()
          ..sort((left, right) => left.index.compareTo(right.index));
    return MetricsSnapshot(
      totalApiCalls: _totalCalls,
      successCalls: _successCalls,
      failureCalls: _failureCalls,
      cacheHits: _cacheHits,
      cacheMisses: _cacheMisses,
      averageLatencyMs: _totalCalls > 0 ? _totalLatencyMs ~/ _totalCalls : 0,
      cacheHitRate: totalRequests > 0 ? _cacheHits / totalRequests : 0,
      perKeyMetrics: summaries,
      recentErrors: recentErrors,
    );
  }

  /// Create a fresh instance for testing isolation.
  /// Tests should call this in setUp() instead of reset().
  static void resetForTesting() {
    instance._totalCalls = 0;
    instance._successCalls = 0;
    instance._failureCalls = 0;
    instance._cacheHits = 0;
    instance._cacheMisses = 0;
    instance._totalLatencyMs = 0;
    instance._callsPerKey.clear();
    instance._errorsPerKey.clear();
    instance._errorLogs.clear();
  }
}

class KeyMetricSummary {
  const KeyMetricSummary({
    required this.index,
    required this.calls,
    required this.errors,
  });

  final int index;
  final int calls;
  final int errors;

  double get errorRate => calls > 0 ? errors / calls : 0;
}

class MetricsSnapshot {
  const MetricsSnapshot({
    required this.totalApiCalls,
    required this.successCalls,
    required this.failureCalls,
    required this.cacheHits,
    required this.cacheMisses,
    required this.averageLatencyMs,
    required this.cacheHitRate,
    this.perKeyMetrics = const <KeyMetricSummary>[],
    this.recentErrors = const <String>[],
  });

  final int totalApiCalls;
  final int successCalls;
  final int failureCalls;
  final int cacheHits;
  final int cacheMisses;
  final int averageLatencyMs;
  final double cacheHitRate;
  final List<KeyMetricSummary> perKeyMetrics;
  final List<String> recentErrors;

  double get successRate =>
      totalApiCalls > 0 ? successCalls / totalApiCalls : 0;

  double get failureRate =>
      totalApiCalls > 0 ? failureCalls / totalApiCalls : 0;
}

File: lachancuocgoi_kotline/app/src/main/java/com/example/lachancuocgoi/Analysis/L3/core/GeminiMetrics.kt
package com.example.lachancuocgoi.Analysis.L3.core

import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicLong

/**
 * Metrics tracker cho Gemini API usage.
 * Thread-safe với atomic operations.
 */
object GeminiMetrics {
    
    private val totalCalls = AtomicInteger(0)
    private val successCalls = AtomicInteger(0)
    private val failureCalls = AtomicInteger(0)
    private val cacheHits = AtomicInteger(0)
    private val cacheMisses = AtomicInteger(0)
    private val totalLatency = AtomicLong(0)
    
    // Per-key metrics
    private val callsPerKey = java.util.concurrent.ConcurrentHashMap<Int, AtomicInteger>()
    private val errorsPerKey = java.util.concurrent.ConcurrentHashMap<Int, AtomicInteger>()
    
    /**
     * Record một API call.
     */
    fun recordCall(success: Boolean, latencyMs: Long, keyIndex: Int = -1) {
        totalCalls.incrementAndGet()
        if (success) {
            successCalls.incrementAndGet()
        } else {
            failureCalls.incrementAndGet()
            if (keyIndex >= 0) {
                errorsPerKey.getOrPut(keyIndex) { AtomicInteger(0) }.incrementAndGet()
            }
        }
        totalLatency.addAndGet(latencyMs)
        if (keyIndex >= 0) {
            callsPerKey.getOrPut(keyIndex) { AtomicInteger(0) }.incrementAndGet()
        }
    }
    
    /**
     * Record cache hit.
     */
    fun recordCacheHit() {
        cacheHits.incrementAndGet()
    }
    
    /**
     * Record cache miss.
     */
    fun recordCacheMiss() {
        cacheMisses.incrementAndGet()
    }
    
    /**
     * Get snapshot của metrics hiện tại.
     */
    fun getSnapshot(): MetricsSnapshot {
        val calls = totalCalls.get()
        val hits = cacheHits.get()
        val misses = cacheMisses.get()
        val totalRequests = hits + misses
        
        // Per-key summary
        val keySummary = callsPerKey.map { (idx, counter) ->
            val errors = errorsPerKey[idx]?.get() ?: 0
            KeyMetricSummary(index = idx, calls = counter.get(), errors = errors)
        }.sortedBy { it.index }
        
        return MetricsSnapshot(
            totalApiCalls = calls,
            successCalls = successCalls.get(),
            failureCalls = failureCalls.get(),
            cacheHits = hits,
            cacheMisses = misses,
            averageLatencyMs = if (calls > 0) totalLatency.get() / calls else 0,
            cacheHitRate = if (totalRequests > 0) hits.toFloat() / totalRequests else 0f,
            perKeyMetrics = keySummary
        )
    }
    
    /**
     * Reset tất cả metrics (for testing).
     */
    fun reset() {
        totalCalls.set(0)
        successCalls.set(0)
        failureCalls.set(0)
        cacheHits.set(0)
        cacheMisses.set(0)
        totalLatency.set(0)
        callsPerKey.clear()
        errorsPerKey.clear()
    }
}

/**
 * Snapshot của metrics tại một thời điểm.
 */
data class KeyMetricSummary(
    val index: Int,
    val calls: Int,
    val errors: Int
) {
    val errorRate: Float get() = if (calls > 0) errors.toFloat() / calls else 0f
}

data class MetricsSnapshot(
    val totalApiCalls: Int,
    val successCalls: Int,
    val failureCalls: Int,
    val cacheHits: Int,
    val cacheMisses: Int,
    val averageLatencyMs: Long,
    val cacheHitRate: Float,
    val perKeyMetrics: List<KeyMetricSummary> = emptyList()
) {
    val successRate: Float
        get() = if (totalApiCalls > 0) successCalls.toFloat() / totalApiCalls else 0f
    
    val failureRate: Float
        get() = if (totalApiCalls > 0) failureCalls.toFloat() / totalApiCalls else 0f
    
    override fun toString(): String {
        val keyLines = if (perKeyMetrics.isNotEmpty()) {
            perKeyMetrics.joinToString("\n") { km ->
                "  Key ${km.index}: ${km.calls} calls, ${km.errors} errors (${(km.errorRate * 100).toInt()}%)"
            }
        } else ""
        return """
            Gemini Metrics:
            - API Calls: $totalApiCalls (✓ $successCalls, ✗ $failureCalls)
            - Success Rate: ${(successRate * 100).toInt()}%
            - Cache: $cacheHits hits, $cacheMisses misses
            - Cache Hit Rate: ${(cacheHitRate * 100).toInt()}%
            - Avg Latency: ${averageLatencyMs}ms
            $keyLines
        """.trimIndent()
    }
}
